src/isomorphic/ReactDebugTool.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactDebugTool
+ */
+
+'use strict';
+
+var ExecutionEnvironment = require('ExecutionEnvironment');
+var ReactPerfDevtool = require('ReactPerfDevtool');
+
+var warning = require('warning');
+
+var eventHandlers = [];
+var handlerDoesThrowForEvent = {};
+
+function emitEvent(handlerFunctionName, arg1, arg2, arg3, arg4, arg5) {
+  if (__DEV__) {
+    eventHandlers.forEach(function(handler) {
+      try {
+        if (handler[handlerFunctionName]) {
+          handler[handlerFunctionName](arg1, arg2, arg3, arg4, arg5);
+        }
+      } catch (e) {
+        warning(
+          !handlerDoesThrowForEvent[handlerFunctionName],
+          'exception thrown by devtool while handling %s: %s',
+          handlerFunctionName,
+          e.message
+        );
+        handlerDoesThrowForEvent[handlerFunctionName] = true;
+      }
+    });
+  }
+}
+
+var ReactDebugTool = {
+  addDevtool(devtool) {
+    eventHandlers.push(devtool);
+  },
+  removeDevtool(devtool) {
+    for (var i = 0; i < eventHandlers.length; i++) {
+      if (eventHandlers[i] === devtool) {
+        eventHandlers.splice(i, 1);
+        i--;
+      }
+    }
+  },
+  onBeginFlushUpdates() {
+    emitEvent('onBeginFlushUpdates');
+  },
+  onStartTimer(internalInstance, timerType) {
+    emitEvent('onStartTimer', internalInstance, timerType);
+  },
+  onEndFlushUpdates() {
+    emitEvent('onEndFlushUpdates');
+  },
+  onStopTimer(internalInstance, timerType) {
```

Comment:
Hi Josh,

Thank you very much for looking through it! I used this approach at first but felt it introduces some undesired bookkeeping in the consuming code. I don’t mind adding it given enough interest but in this case I think we’re covered well by the guarantee that the scopes are never nested.

One of the goals of this refactoring is to stop relying on stacked calls, and explicitly forcing `stop()` calls to *match exactly* the preceding `start()` call without keeping track of the stack gives even an even stronger guarantee than a scope token unless I am missing something.